File: src/net/GOOM.TR.MyRetail.NET/GOOM.TR.MyRetail.NET/Repos/PriceRepo.cs
using GOOM.TR.MyRetail.NET.Models;
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;
using MongoDB.Driver;
using System.Threading.Tasks;
using MongoDB.Bson.Serialization.IdGenerators;

namespace GOOM.TR.MyRetail.NET.Repos
{
    public interface IPriceRepo
    {
        Task<Price> GetPriceByProductIdAsync(long productId);
        Task<Price> SetProductPriceAsync(long productId, Price price);
        Task<bool> DeleteProductPriceAsync(long productId);
    }

    public class PriceRepo : IPriceRepo
    {
        private readonly IMongoCollection<DbPrice> prices;

        public PriceRepo(IPriceStoreDatabaseSettings settings)
        {
            var client = new MongoClient(settings.ConnectionString);
            var database = client.GetDatabase(settings.DatabaseName);

            prices = database.GetCollection<DbPrice>(settings.PricesCollectionName);
        }
        
        public async Task<Price> GetPriceByProductIdAsync(long productId)
        {
            var dbPrice = (await prices.FindAsync(x => x.ProductId == productId)).SingleOrDefault();
            return dbPrice == null
                ? null
                : new Price
                {
                    CurrencyCode = dbPrice.CurrencyCode,
                    Value =  dbPrice.Value
                };
        }

        public async Task<Price> SetProductPriceAsync(long productId, Price price)
        {
            var result = await prices.ReplaceOneAsync(x => x.ProductId == productId,
                new DbPrice()
                {
                    ProductId = productId,
                    CurrencyCode = price.CurrencyCode, 
                    Value = price.Value
                }, new ReplaceOptions() { IsUpsert = true});
            
            return result.IsAcknowledged ? price : null;
        }

        public async Task<bool> DeleteProductPriceAsync(long productId)
        {
            var result = await prices.DeleteOneAsync(x => x.ProductId == productId);

            return result.DeletedCount == 1;
        }
    }

    public class DbPrice
    {
        [BsonIgnoreIfDefault]
        [BsonRepresentation(BsonType.ObjectId)]
        [BsonId(IdGenerator = typeof(StringObjectIdGenerator))]
        public string Id { get; set; }

        public long ProductId { get; set; }

        public decimal Value { get; set; }

        public string CurrencyCode { get; set; }
    }

    public class PriceStoreDatabaseSettings : IPriceStoreDatabaseSettings
    {
        public string PricesCollectionName { get; set; }
        public string ConnectionString { get; set; }
        public string DatabaseName { get; set; }
    }

    public interface IPriceStoreDatabaseSettings
    {
        string PricesCollectionName { get; set; }
        string ConnectionString { get; set; }
        string DatabaseName { get; set; }
    }
}

File: src/java/src/main/java/com/goom/tr/myretail/repos/PriceRepo.java
package com.goom.tr.myretail.repos;

import com.goom.tr.myretail.models.Price;
import com.goom.tr.myretail.repos.mongo.PricesRepository;
import com.goom.tr.myretail.repos.mongo.models.DbPrice;

public class PriceRepo implements IPriceRepo {
	private final PricesRepository _pricesRepository;
	
	public PriceRepo(PricesRepository pricesRepository) {
		_pricesRepository = pricesRepository;
	}
	
	@Override
	public Price GetPriceByProductId(long productId) {
		DbPrice dbPrice = _pricesRepository.findByProductId(productId);
		if(dbPrice == null) return null;
		
		var price = new Price();
		price.CurrencyCode = dbPrice.currencyCode;
		price.Value = dbPrice.value;
		return price;
	}

	@Override
	public Price SetProductPrice(long productId, Price price) {

		DbPrice dbPrice = _pricesRepository.findByProductId(productId);
		if(dbPrice == null) { 
			dbPrice = new DbPrice(); 
			dbPrice.productId = productId;
		}
		dbPrice.currencyCode = price.CurrencyCode;
		dbPrice.value = price.Value;
		
		
		_pricesRepository.save(dbPrice);
		
		return price;
	}

	@Override
	public boolean DeleteProductPrice(long productId) {
		DbPrice dbPrice = _pricesRepository.findByProductId(productId);
		if(dbPrice == null) return false;
		
		_pricesRepository.delete(dbPrice);
		return true;
	}
}
